File: Els_kom/Forms/Els_kom_Main_SettingsForm.cs
// Copyright (c) 2014-2019, Els_kom org.
// https://github.com/Elskom/
// All rights reserved.
// license: MIT, see LICENSE for more details.

namespace Els_kom.Forms
{
    using System;
    using System.Threading.Tasks;
    using System.Windows.Forms;
    using Elskom.Generic.Libs;

    internal partial class SettingsForm : Form
    {
        internal SettingsForm() => this.InitializeComponent();

        internal static string Label1 { get; private set; } = "0";

        private static void OpenPluginSettings(Form plugsettingfrm, IWin32Window owner)
            => plugsettingfrm.ShowDialog(owner);

        private void SettingsForm_Load(object sender, EventArgs e)
        {
            // This new member is to ensure the data from
            // the control load function does not try to get the data
            // from the Settings XMLObject in Designer mode
            // (which contains no instance there anyway) that causes
            // the Designer to fail to load the form the control is on.
            this.SettingsControl1.InitControl();
            Label1 = "1";
        }

        private void SettingsForm_FormClosing(object sender, FormClosingEventArgs e)
        {
            Label1 = "0";
            this.SettingsControl1.SaveSettings();
        }

        private void SettingsControl1_OpenPluginsForm(object sender, EventArgs e)
        {
            using (var pluginsForm = new PluginsForm())
            {
                pluginsForm.ShowDialog();
            }
        }

        private void SettingsControl1_OpenPluginsSettings(object sender, EventArgs e)
        {
            for (var i = 0; i < this.SettingsControl1.ListView1.SelectedItems.Count; i++)
            {
                var selitem = this.SettingsControl1.ListView1.SelectedItems[i];
                foreach (var callbackplugin in KOMStream.Callbackplugins)
                {
                    if (callbackplugin.PluginName.Equals(selitem.Text))
                    {
                        var plugsettingfrm = callbackplugin.SettingsWindow;
                        plugsettingfrm.Icon = Icons.FormIcon;
                        Task.Run(
                            () => OpenPluginSettings(
                                plugsettingfrm,
                                callbackplugin.ShowModal ? this : null));
                    }
                }
            }
        }
    }
}

File: PC/Els_kom_Main_SettingsForm.cpp
#include "LoadResources.hpp"
#include "resource.hpp"
#include "Els_kom_Main_SettingsForm.hpp"

using namespace LoadResources;

namespace Els_kom {
  namespace Forms {
    SettingsForm::SettingsForm(void)
    {
      InitializeComponent();
      //
      //TODO: Add the constructor code here
      //
    }

    SettingsForm::~SettingsForm()
    {
      if (components)
      {
        delete components;
      }
    }

    System::Void SettingsForm::SettingsForm_Load(System::Object^  sender, System::EventArgs^  e) {
      this->Icon = GetIconResource(IDI_MAINICON);
    }

    System::Void SettingsForm::SettingsForm_FormClosing(System::Object^  sender, System::Windows::Forms::FormClosingEventArgs^  e) {
      this->settingsControl1->SaveSettings();
    }

    System::Void SettingsForm::settingsControl1_CloseForm(System::Object^  sender, System::EventArgs^  e) {
      this->Close();
    }
  }
}
